core/drivers/clk/clk_dt.c

```diff
@@ -0,0 +1,395 @@
+
+static struct clk *clk_dt_get_by_idx_prop(const char *prop_name,
+					  const void *fdt, int nodeoffset,
+					  unsigned int clk_idx)
+{
+	int len = 0;
+	int idx = 0;
+	int idx32 = 0;
+	int clock_cells = 0;
+	uint32_t phandle = 0;
+	const uint32_t *prop = NULL;
+	struct clk_dt_provider *prv = NULL;
+
+	prop = fdt_getprop(fdt, nodeoffset, prop_name, &len);
+	if (!prop)
+		return NULL;
+
+	while (idx < len) {
+		idx32 = idx / sizeof(uint32_t);
+		phandle = fdt32_to_cpu(prop[idx32]);
+
+		prv = clk_get_provider_by_phandle(phandle);
+		if (!prv)
+			return NULL;
+
+		clock_cells = prv->clock_cells;
+		if (clk_idx) {
+			clk_idx--;
+			idx += sizeof(phandle) + clock_cells * sizeof(uint32_t);
+			continue;
+		}
+
+		return clk_dt_get_from_provider(prv, clock_cells, &prop[idx32]);
+	}
+
+	return NULL;
+}
+
+struct clk *clk_dt_get_by_idx(const void *fdt, int nodeoffset,
+			      unsigned int clk_idx)
+{
+	return clk_dt_get_by_idx_prop("clocks", fdt, nodeoffset, clk_idx);
+}
+
+static const struct clk_driver *clk_get_compatible_driver(const char *compat)
+{
+	const struct dt_driver *drv = NULL;
+	const struct dt_device_match *dm = NULL;
+	const struct clk_driver *clk_drv = NULL;
+
+	for_each_dt_driver(drv) {
+		if (drv->type != DT_DRIVER_CLK)
+			continue;
+
+		clk_drv = (const struct clk_driver *)drv->driver;
+		for (dm = drv->match_table; dm && dm->compatible; dm++) {
+			if (strcmp(dm->compatible, compat) == 0)
+				return clk_drv;
+		}
+	}
+
+	return NULL;
+}
+
+/* Recursively called from parse_clock_property() */
+static TEE_Result clk_probe_clock_provider_node(const void *fdt, int node);
+
+static TEE_Result parse_clock_property(const void *fdt, int node)
+{
+	int len = 0;
+	int idx = 0;
+	int parent_node = 0;
+	int clock_cells = 0;
+	uint32_t phandle = 0;
+	const uint32_t *prop = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	prop = fdt_getprop(fdt, node, "clocks", &len);
+	if (!prop)
+		return TEE_SUCCESS;
+
+	len /= sizeof(uint32_t);
+	while (idx < len) {
+		phandle = fdt32_to_cpu(prop[idx]);
+
+		parent_node = fdt_node_offset_by_phandle(fdt, phandle);
+		if (parent_node < 0)
+			return TEE_ERROR_GENERIC;
+
+		/* Parent setup should not fail or clock won't be available */
+		res = clk_probe_clock_provider_node(fdt, parent_node);
+		if (res)
+			panic("Failed to setup parent clock");
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
+		idx += 1 + clock_cells;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_dt_node_clock_setup_driver(const void *fdt, int node)
+{
+	int idx = 0;
+	int len = 0;
+	int count = 0;
+	const char *compat = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const struct clk_driver *clk_drv = NULL;
+
+	count = fdt_stringlist_count(fdt, node, "compatible");
+	if (count < 0)
+		return TEE_ERROR_GENERIC;
+
+	for (idx = 0; idx < count; idx++) {
+		compat = fdt_stringlist_get(fdt, node, "compatible", idx,
+					    &len);
+		if (!compat)
+			return TEE_ERROR_GENERIC;
+
+		clk_drv = clk_get_compatible_driver(compat);
```

Comment:
Wouldn't `dt_find_compatible_driver(fdt, node)` give the same result?

Reply:
Let's say that your node as the following compatible in your node: `compatible = "compat1", "compat2"`
If the driver array contains the driver for compat2 before the driver for compat1 then the order of compatible would not be honored. I guess this should probably be fixed in `dt_find_compatible_driver`. @etienne-lms weren't you talking about that in some comment ?

Reply:
Yes, `dt_find_compatible_driver()` does not take into account the ordering in property compatible.
Maybe `clk_get_compatible_driver()` could be changed to use a fixed implementation of `dt_find_compatible_driver()`.